### PR TITLE
BDPT work

### PR DESCRIPTION
**Pull request by aytekaman**

### CODE REVIEW

File: src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp [hunk 1]
```diff
@@ -78,6 +81,19 @@ namespace
             const ShadingPoint&     shading_point,
             ShadingComponents&      radiance) override      // output radiance, in W.sr^-1.m^-2
         {
+            PathVisitor light_path_visitor;
+            VolumeVisitor volume_visitor;
+
+            PathTracer<PathVisitor, VolumeVisitor, false> light_path_tracer(    // false = not adjoint
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp]: For light tracing, you'll want to set `Adjoint` to `true`. You can find an extended treatment of adjoint tracing in Veach's thesis.

**Comment by dictoon** [comment]
It's a good start.

Regarding how to trace the light path, you can find inspiration in the Light Tracer:
https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/renderer/kernel/lighting/lighttracing/lighttracingsamplegenerator.cpp#L570-L786

**Comment by dictoon** [comment]
I suggest to keep this PR open so that you can progressively add commit until a first working implementation. How does that sound?

**Comment by aytekaman** [comment]
Sounds like a plan:)

File: src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.h [hunk 2]
```diff
@@ -48,7 +49,9 @@ class BDPTLightingEngineFactory
 {
   public:
     // Constructor.
-    explicit BDPTLightingEngineFactory(const ParamArray& params);
+    explicit BDPTLightingEngineFactory(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.h]: You can safely remove `explicit` now that the constructor takes more than one argument. I just wish C++ had explicit constructors by default :)

Comment (aytekaman) [reply src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.h]: Ah, you are right. Will do it.

File: src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.h [hunk 3]
```diff
@@ -33,8 +33,13 @@
 #include "renderer/kernel/lighting/ilightingengine.h"
 #include "renderer/utility/paramarray.h"
 
+// appleseed.foundation headers.
+#include "foundation/math/population.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.h]: It doesn't look like you need this header here? Maybe only in the .cpp file?

**Pull request merged**